crates/subspace-networking/src/protocols/connected_peers/tests.rs

```diff
@@ -0,0 +1,272 @@
+                panic!("No new peers requests.");
+            }
+        }
+    }
+}
+
+#[tokio::test()]
+async fn test_target_connected_peer_limit_number() {
+    let decision_timeout = Duration::from_millis(500);
+    let large_delay = Duration::from_millis(2000);
+    let target_connected_peers = 1;
+
+    let mut peer1 = new_ephemeral(
+        decision_timeout,
+        Behaviour::<ConnectedPeersInstance>::new(Config {
+            decision_timeout,
+            target_connected_peers,
+            ..Default::default()
+        }),
+    );
+
+    let mut peer2 = new_ephemeral(
+        decision_timeout,
+        Behaviour::<ConnectedPeersInstance>::new(Config {
+            target_connected_peers,
+            decision_timeout,
+            ..Default::default()
+        }),
+    );
+
+    let mut peer3 = new_ephemeral(
+        decision_timeout,
+        Behaviour::<ConnectedPeersInstance>::new(Config {
+            target_connected_peers,
+            decision_timeout,
+            ..Default::default()
+        }),
+    );
+
+    peer2.listen().await;
+    peer3.listen().await;
+
+    peer1.connect(&mut peer2).await;
+    peer1.connect(&mut peer3).await;
+
+    peer2.connect(&mut peer3).await;
+
+    peer1
+        .behaviour_mut()
+        .update_keep_alive_status(*peer2.local_peer_id(), true);
+    peer1
+        .behaviour_mut()
+        .update_keep_alive_status(*peer3.local_peer_id(), true);
+
+    peer2
+        .behaviour_mut()
+        .update_keep_alive_status(*peer1.local_peer_id(), true);
+    peer2
+        .behaviour_mut()
+        .update_keep_alive_status(*peer3.local_peer_id(), true);
+
+    peer3
+        .behaviour_mut()
+        .update_keep_alive_status(*peer1.local_peer_id(), true);
+    peer3
+        .behaviour_mut()
+        .update_keep_alive_status(*peer2.local_peer_id(), true);
+
+    loop {
+        select! {
+            _ = peer1.next_swarm_event().fuse() => {},
+            _ = peer2.next_swarm_event().fuse() => {},
+            _ = peer3.next_swarm_event().fuse() => {},
+            _ = sleep(large_delay).fuse() => {
+                break;
+            }
+        }
+    }
+
+    assert!(peer1.is_connected(peer2.local_peer_id()));
+    assert!(!peer1.is_connected(peer3.local_peer_id()));
+
+    assert!(peer2.is_connected(peer1.local_peer_id()));
+    assert!(!peer2.is_connected(peer3.local_peer_id()));
+
+    assert!(!peer3.is_connected(peer1.local_peer_id()));
+    assert!(!peer2.is_connected(peer2.local_peer_id()));
+}
+
+fn new_ephemeral<NB: NetworkBehaviour>(connection_timeout: Duration, behaviour: NB) -> Swarm<NB> {
+    let identity = Keypair::generate_ed25519();
+    let peer_id = PeerId::from(identity.public());
+
+    let transport = MemoryTransport::default()
+        .or_transport(libp2p::tcp::tokio::Transport::default())
+        .upgrade(Version::V1)
+        .authenticate(PlainText2Config {
+            local_public_key: identity.public(),
+        })
```

Comment:
Why plain text though, how do we benefit from it? Crypto cost should be minimal.

Reply:
I initially removed that but `swarm-test` crate has "test-only" quality and it doesn't work without it.